runtime/lua/vim/cache.lua

```diff
@@ -0,0 +1,481 @@
+  end
+
+  -- Checks if the given paths contain the top-level module.
+  -- If so, it tries to find the module path for the given module name.
+  ---@param paths string[]
+  ---@return string? modpath, CacheHash? hash
+  ---@private
+  local function _find(paths)
+    for _, path in ipairs(paths) do
+      if M.lsmod(path)[topmod] then
+        for _, pattern in ipairs(patterns) do
+          local modpath = path .. pattern
+          M.stats.find.stat = (M.stats.find.stat or 0) + 1
+          local hash = uv.fs_stat(modpath)
+          if hash then
+            return modpath, hash
+          end
+        end
+      end
+    end
+  end
+
+  ---@type string?, CacheHash?
+  local modpath, hash
+
+  -- always check the rtp first
+  if opts.rtp ~= false then
+    modpath, hash = _find(Cache._rtp or {})
+    if not modpath then
+      local rtp, updated = Cache.get_rtp()
+      if updated then
+        modpath, hash = _find(rtp)
+      end
+    end
+  end
+
+  -- check any additional paths
+  if (not modpath) and opts.paths then
+    modpath, hash = _find(opts.paths)
+  end
+
+  M.track('find', start)
+  if modpath then
+    return modpath, hash
+  end
+  -- module not found
+  M.stats.find.not_found = M.stats.find.not_found + 1
+end
+
+--- Resets the topmods cache for the path
+---@param path string path to reset
+function M.reset(path)
+  Cache._indexed[Cache.normalize(path)] = nil
+end
+
+--- Enables the cache:
+--- * override loadfile
+--- * adds the lua loader
+--- * adds the libs loader
+--- * remove the Neovim loader
+function M.enable()
+  if M.enabled then
+    return
+  end
+  M.enabled = true
+  vim.fn.mkdir(vim.fn.fnamemodify(M.path, ':p'), 'p')
+  -- selene: allow(global_usage)
+  _G.loadfile = Cache.loadfile
+  -- add lua loader
+  table.insert(package.loaders, 2, Cache.loader)
+  -- add libs loader
+  table.insert(package.loaders, 3, Cache.loader_lib)
+  -- remove Neovim loader
+  for l, loader in ipairs(package.loaders) do
+    if loader == vim._load_package then
+      table.remove(package.loaders, l)
+      break
+    end
+  end
+  -- TODO: add an autocmd on BufWritePost that checks if its in a /lua folder
+  -- if thats the case, then reset the plugin path.
+  -- This will make sure we can properly load new top-level lua modules
+end
+
+--- Disables the cache:
+--- * removes the cache loaders
+--- * adds the Neovim loader
+function M.disable()
+  if not M.enabled then
+    return
+  end
+  M.enabled = false
+  -- selene: allow(global_usage)
+  _G.loadfile = Cache._loadfile
+  ---@diagnostic disable-next-line: no-unknown
+  for l, loader in ipairs(package.loaders) do
+    if loader == Cache.loader or loader == Cache.loader_lib then
+      table.remove(package.loaders, l)
+    end
+  end
+  table.insert(package.loaders, 2, vim._load_package)
+end
+
+--- Return the top-level `/lua/*` modules for this path
+---@param path string path to check for top-level lua modules
+---@return string[]
+function M.lsmod(path)
```

Comment:
Can this be made private?

Reply:
used by lazy, so I don't think it's a good idea to make this private